pkg/sql/schemaexpr/check_constraint_test.go

```diff
@@ -0,0 +1,193 @@
+
+		// Use unique default names when there is no custom name.
+		{"", "a", true, "a", "check_a"},
+		{"", "a", true, "a", "check_a1"},
+		{"", "a", true, "a", "check_a2"},
+		{"", "a AND b = 0", true, "a AND (b = 0)", "check_a_b"},
+		{"", "a AND b = 1", true, "a AND (b = 1)", "check_a_b1"},
+		{"", "a AND b = 1", true, "a AND (b = 1)", "check_a_b2"},
+
+		// Respect that "check_a3" has been marked, so the next check constraint
+		// with "a" should be "check_a4".
+		{"", "a", true, "a", "check_a4"},
+		{"", "a", true, "a", "check_a5"},
+
+		// Allow expressions that result in a bool.
+		{"ck", "a", true, "a", "ck"},
+		{"ck", "b = 0", true, "b = 0", "ck"},
+		{"ck", "a AND b = 0", true, "a AND (b = 0)", "ck"},
+		{"ck", "a IS NULL", true, "a IS NULL", "ck"},
+		{"ck", "b IN (1, 2)", true, "b IN (1:::INT8, 2:::INT8)", "ck"},
+
+		// Allow immutable functions.
+		{"ck", "abs(b) > 0", true, "abs(b) > 0", "ck"},
+		{"ck", "c || c = 'foofoo'", true, "(c || c) = 'foofoo'", "ck"},
+		{"ck", "lower(c) = 'bar'", true, "lower(c) = 'bar'", "ck"},
+
+		// Allow mutable functions.
+		{"ck", "b > random()", true, "b > random()", "ck"},
+
+		// Disallow references to columns not in the table.
+		{"", "d", false, "", ""},
+		{"", "t.a", false, "", ""},
+
+		// Disallow expressions that do not result in a bool.
+		{"", "b", false, "", ""},
+		{"", "abs(b)", false, "", ""},
+		{"", "lower(c)", false, "", ""},
+
+		// Disallow subqueries.
+		{"", "exists(select 1)", false, "", ""},
+		{"", "b IN (select 1)", false, "", ""},
+
+		// Disallow aggregate, window, and set returning functions.
+		{"", "sum(b) > 10", false, "", ""},
+		{"", "row_number() OVER () > 1", false, "", ""},
+		{"", "generate_series(1, 1) > 2", false, "", ""},
+
+		// Dequalify column names.
+		{"ck", "bar.a", true, "a", "ck"},
+		{"ck", "foo.bar.a", true, "a", "ck"},
+		{"ck", "bar.b = 0", true, "b = 0", "ck"},
+		{"ck", "foo.bar.b = 0", true, "b = 0", "ck"},
+		{"ck", "bar.a AND foo.bar.b = 0", true, "a AND (b = 0)", "ck"},
+	}
+
+	for _, d := range testData {
+		t.Run(d.expr, func(t *testing.T) {
+			expr, err := parser.ParseExpr(d.expr)
+			if err != nil {
+				t.Fatalf("%s: unexpected error: %s", d.expr, err)
+			}
+
+			ckDef := &tree.CheckConstraintTableDef{Name: tree.Name(d.name), Expr: expr}
+
+			ck, err := builder.Build(ckDef)
+
+			if !d.expectedValid {
+				if err == nil {
+					t.Fatalf("%s: expected invalid expression, but was valid", d.expr)
+				}
+				// The input expression is invalid so there is no need to check
+				// the output ck.
+				return
+			}
+
+			if err != nil {
+				t.Fatalf("%s: expected valid expression, but found error: %s", d.expr, err)
+			}
+
+			if ck.Name != d.expectedName || ck.Expr != d.expectedExpr {
+				t.Errorf(
+					`%s: expected "%s CHECK %s", got "%s CHECK %s"`,
+					d.expr,
+					d.expectedName,
+					d.expectedExpr,
+					ck.Name,
+					ck.Expr,
+				)
+			}
+		})
+	}
+}
+
+func TestCheckConstraintBuilder_DefaultName(t *testing.T) {
```

Comment:
Do you think this test is somewhat brittle? Or we aren't changing how we generate names for default columns in the future

Reply:
In my mind, a "brittle" test is one that breaks when the implementation details change but the API "contract" (in this case input/output) does not. I don't think this is brittle because it is guaranteeing a specific output based on a specific input, but doesn't assert how that output is generated.

If we did want to change this "contract" in the future (AFAIK we have no plans to) then updating the tests to guarantee the new contract seems like a valuable step to me.

Reply:
I see -- I wasn't sure there was a set contract on how the constraint names will be formatted.